### ISSUE TITLE
missing reflink for linkType

### ISSUE THREAD

**Issue by KyleSanderson** [cross-seed/cross-seed#421]
Microsoft exposed the venerable deduplication facility to NTFS userspace over 10 years ago in Server 2012, with Linux picking up the feature in BTRFS shortly thereafter (reflink) around 3.6. XFS around 4.9 added support for reflink as well, with them being declared feature stable long ago. Other filesystems support this (namely, APFS), marking the holy trinity for operating system support.

When data is soft linked, the parent inode is shared, along with the data underneath.
When data is hard linked, the parent inode is unique, but all extents are shared.
When data is ref linked, the extents are marked as the same in both locations, but allow for writes to change the contents without impacting the other files (which, transparently breaks the sharing relationship). This way even on a partial match, you can still achieve significant space savings.

**Comment by zakkarry** [comment]
> Reflinking would solve my problems:
> 
>     * symlinks break when the source is moved

Yes, they do, but for some people this, and not having a single device/partition/array is why they use them. When the original torrent with all of the download associated with it is deleted - they may no longer want the additional cross-seeds to stick around. They error and thus show them "Hey delete these" - it's a workflow I can't fault them for as I have my own plugin for deluge to point out any error'd torrents duplicates to see if I want to trump or delete them. Just a different mechanism to do that with symlinks.

There are benefits, however, we recommend hardlinks over symlinks if they are possible with the user's setup.

> 
>     * hardlinks cannot be used across zfs datasets
I see, I've heard this mentioned in the past. Is there a particular reason you can't hardlink to the same dataset? I'm not familiar with how zfs works beyond broadly, so you'll have to enlighten me as to why this is an issue. As I remember, your datasets are merged into a single volume or something? Do I have that right? Those datasets are the vdevs?

Does zfs not have logic to hardlink to the same dataset when you make a hardline on the volume? That seems like a trivial thing for them to integrate given the complexity of zfs as I understand it already.

Sorry for my ignorance, just trying to understand what you can and can't do and get a little context. The unfortunate side of this is that if this is strictly a ZFS requirement, there isn't a huge populous of users who would have ZFS.


>     * hardlinks are not ideal for partials, where a small delta in a file may exist in a cross seed torrent, but changing the existing file in my library is not wanted.

I'm not sure if you're misunderstanding what a partial match is, or if I'm misunderstanding what you're trying to say. It is exceedingly rare, if you have even the top of the low-tier trackers, to find files that are different across trackers for the same release that also match. These files almost always would be corrupted, or possibly a .torrent file creation problem by the uploader, and not something you should _WANT_ to seed further.

Generally, partial matching in the context cross-seed refers to is when you have a torrent with maybe an .nfo or .srt file, or a season pack that you find had one episode repacked and reuploaded as a "new" pack. Partial does not refer to 5KB out of a 10KB file different and matching the file anyway. When this potentially happens (as I said - very rarely in the vast majority of cases) it is usually deleted once a user reports the torrent anyway.

The way cross-seed currently works, for the majority of matches we aren't exceedingly sure the risk of the files differing is low, we do a recheck. Currently, that will leave the torrents paused. When you see these torrents sitting at <100% paused, you can look at if there is only an nfo file missing and 99% of the main file - in which case due to [piece validation hashing](https://www.cross-seed.org/docs/v6-migration#partial-matching) - see the note at the top of this section - this is fine to resume and will not rarely modify the original media file - but if there is a TV show at 80%, the match is a false positive and one of the two is almost certainly corrupted in some way. 

While I understand it "would be cool", I'm sort of in a position of wanting to understand what shortcomings you currently face in a real use case and why it currently isn't possible, I think my lack of knowledge of ZFS might be a contributing factor to this. Frankly, it may not be overly complex to support them, but if it turns out to be something we have to walk users through setting up to work - that support burden alone would be my main concern. A large majority of the current support load is already me being tasked with helping users set up volumes and getting linking to function. I'd expect adding a file system and a new potential linking setup for users to "want" wouldn't ease that load.

That's not to say we can't or won't do this. At a glance, it seems like something we could support. I'm just a curious person by nature and want to understand.

My only real concern, having not investigated it much, that is causing hesitation comes from having to provide support after it is implemented.

**Comment by mmgoodnow** [comment]
I think reflink support would be a reasonable thing to add. It adds another unique linking behavior that could be desirable to some (seeding as many things as possible with no regard for HDD space)

**Comment by zakkarry** [comment]
> I think reflink support would be a reasonable thing to add. It adds another unique linking behavior that could be desirable to some (seeding as many things as possible with no regard for HDD space)

As I said, upon initially looking into it there was nothing inherently problematic. But if there is an influx of support issues for reflinks and ZFS problems, what do we do...."Sorry, go find someone else" ?

Hardlinks are already the most common issue, and linking configurations/volumes in general.

**Comment by zakkarry** [comment]
Oh, thanks for giving your opinion on what I may or may not have said without reading it at all.

My primary concern wasn't about adding it to the code.

**Comment by zakkarry** [comment]
Here's another wall of text for you since you liked the first one so much.

>Every single filesystem supports reflinks

>ext4, does not support reflinks

Right... that tracks right there.

>Now you've tricked me into responding to a completely unrelated statement, as opposed to the feature at hand.

No, you can't help but be an asshole, and the community at large knows it/you very well by reputation alone. Despite claiming that I was the instigator here for your response, you failed to address what I repeatedly voiced as my concern. The point is that users of systems like TrueNAS, who we all know have no real viable support channels to begin with, or even just regular ZFS setups seeking support with cross-seed will find more trouble than other linking would have.

The number of users of that currently is small, but growing, as ZFS gets more widespread adoption. The number of people who actively do support on our discord who currently have the requisite knowledge of either is basically nill.

The problem is adding a feature and telling the user "Good Luck, hope you figure it out" is not an image I _want_ to portray - as I'm the main one on the team doing the user support. That's why I asked questions in my post, so I would get an idea of what sort of scenario this would work for. Instead you argue that "NO CUZ REASONS" bullshit. 

Whether or not it's beneficial or not is not even something I contested. Every feature can be beneficial to somebody. 

Once again you are too eager to try and prove to everyone that you are the biggest and baddest genius to get online., when in fact you just look pompous and arrogant. "There is no debate about this." You've done that on this repo before, and here you are champing at the bit to get into another argument here.

>Saying that there's no partials, or that they are rare, is just flat out not knowing or having visibility into markets.

I would bet money that I've been doing this shit longer than you Kyle. My visibility is guaranteed to be as good as or better than yours. You can take that to the bank.

Furthermore, I'd love to hear your examples of valid cross-seeds of the same release that differ in their actual data. Please explain how a groups episode you downloaded on one tracker turned out to be different on another without being corrupted.

If this is common, you need to find better trackers.
